xds/src/main/java/io/grpc/xds/AbstractXdsClient.java

```diff
@@ -0,0 +1,760 @@
+    // Client initiated requests start response nonce with empty string.
+    // A nonce is used to indicate the specific DiscoveryResponse each DiscoveryRequest
+    // corresponds to.
+    // A nonce becomes stale following a newer nonce being presented to the client in a
+    // DiscoveryResponse.
+    private String ldsRespNonce = "";
+    private String rdsRespNonce = "";
+    private String cdsRespNonce = "";
+    private String edsRespNonce = "";
+
+    abstract void start();
+
+    abstract void sendError(Exception error);
+
+    /**
+     * Sends a discovery request with the given {@code versionInfo}, {@code nonce} and
+     * {@code errorDetail}. Used for reacting to a specific discovery response. For
+     * client-initiated discovery requests, use {@link
+     * #sendDiscoveryRequest(ResourceType, Collection)}.
+     */
+    abstract void sendDiscoveryRequest(ResourceType type, String versionInfo,
+        Collection<String> resources, String nonce, @Nullable String errorDetail);
+
+    /**
+     * Sends a client-initiated discovery request.
+     */
+    final void sendDiscoveryRequest(ResourceType type, Collection<String> resources) {
+      String nonce;
+      switch (type) {
+        case LDS:
+          nonce = ldsRespNonce;
+          break;
+        case RDS:
+          nonce = rdsRespNonce;
+          break;
+        case CDS:
+          nonce = cdsRespNonce;
+          break;
+        case EDS:
+          nonce = edsRespNonce;
+          break;
+        case UNKNOWN:
+        default:
+          throw new AssertionError("Unknown resource type: " + type);
+      }
+      sendDiscoveryRequest(type, getCurrentVersion(type), resources, nonce, null);
+    }
+
+    final void handleRpcResponse(
+        ResourceType type, String versionInfo, List<Any> resources, String nonce) {
+      if (closed) {
+        return;
+      }
+      responseReceived = true;
+      // Nonce in each response is echoed back in the following ACK/NACK request. It is
+      // used for management server to identify which response the client is ACKing/NACking.
+      // To avoid confusion, client-initiated requests will always use the nonce in
+      // most recently received responses of each resource type.
+      switch (type) {
+        case LDS:
+          ldsRespNonce = nonce;
+          handleLdsResponse(versionInfo, resources, nonce);
+          break;
+        case RDS:
+          rdsRespNonce = nonce;
+          handleRdsResponse(versionInfo, resources, nonce);
+          break;
+        case CDS:
+          cdsRespNonce = nonce;
+          handleCdsResponse(versionInfo, resources, nonce);
+          break;
+        case EDS:
+          edsRespNonce = nonce;
+          handleEdsResponse(versionInfo, resources, nonce);
+          break;
+        case UNKNOWN:
+        default:
+          logger.log(XdsLogLevel.WARNING, "Ignore an unknown type of DiscoveryResponse");
```

Comment:
Provide the type name?

Reply:
Hmm, the enum value doesn't carry the raw type name (aka, typeUrl), this method is already using abstracted information. The real type URL should have been logged in the raw message, so it should be good enough to just warn with the the unknownness.

Reply:
Yeah, we can't print the actual type name.